### PR TITLE
BASE BRANCH: Internal compilation framework

### LINKED ISSUE #298: Remove Support/Dependency on _ANY_ development framework
Currently, we lean heavily on the user's local machine having either Foundry or Hardhat installed. We also require that the directory that aderyn is being run on is a Foundry or Hardhat project. Foundry has good support and aderyn works very well with it, Hardhat less so.

This issue is to track the goal of not requiring any development framework. Aderyn should just *work* on a directory full of solidity files (so long as they compile).

This removes much of the headache of continually supporting multiple frameworks, as well as negating issues with custom Foudnry/Hardhat settings on the local machine/directory.

### PR DESCRIPTION
**Pull request by alexroan**
Resolves #298 

Tasks:
* Remove dependency on Foundry or Hardhat being installed locally.  ✅ 
* Remove dependency on the workspace being a Foundry or Hardhat project.  ✅ 
* Do not require any prerequisite steps like installing `solc` locally before running aderyn. ✅ 
* **Maintain feature parity with current version of aderyn**
* Version lock foundry-compilers dependency  ✅ 
* Handle Remappings   ✅ 
* ~Explore caching mechanisms (we no longer can take forge's help)~ (After merge to dev)
* Add `--icf` flag to opt-in ✅ 

Completed Feature parity
* Markdown report ✅
* JSON report ✅
* Sarif report ✅
* Auditor mode ✅
* CI tests for all reports (both ICF and legacy) ✅ 

Feature parity caveats
* Works on foundry project as-is ✅

* Works on hardhat project 
a) In ICF mode must pass  `--src contracts/` and have `remappings.txt` in project root) ✅
b) In legacy mode (a.k.a without the `--icf` flag) ther is NO support.  ✅

* Single file mode - TODO

Tentative (✅)

~Browse commonly known files and environment vars like `foundry.toml` if present,  to find out scope (by looking at src attribute value in foundry profile)~ We use foundry's official config crate to achieve this

Breaking changes  
* `requirements` or `requirements.txt` needs to be present in the project root if needed (not needed for foundry)

New changes
*  Introducing `--src` which is a comma-separated list of paths relative to the project root. Aderyn will only look inside the sources and follow `--scope` and `--exclude` (if set)

* This `--src` file only works in `--icf` mode

### CODE REVIEW

**Comment by alexroan** [comment]
@TilakMaddy, once this branch CI is fixed, use this branch like we currently use `dev`. I.e., Create branches and PRs that can be reviewed and merged into this one individually.

I'll periodically resolve conflicts from `dev` whilst you're working on this long-term branch.

Once we check off all the tasks, we can merge this one into dev :)

**Comment by alexroan** [comment]
@TilakMaddy I'm going to resolve these conflicts (I'm letting you know in case you've started working on it)

**Comment by TilakMaddy** [comment]
I'm currently looking into how the error can be reproduced, Alex. Let me know if you find out : )

**Comment by TilakMaddy** [comment]
The import error is because the libraries are git submodules which don't exist unless explicitly cloned recursively

**Comment by TilakMaddy** [comment]
@alexroan Report checking has been sorted out. Only when it comes to tests in CI, it fails for the following reason 

```bash
thread 'detect::experimental::sibling::sibling_demo_tests::test_siblings_by_loading_contract_directly' panicked at aderyn_core/src/detect/detector.rs:292:10:
called `Result::unwrap()` on an `Err` value: SvmError(IoError(Os { code: 26, kind: ExecutableFileBusy, message: "Text file busy" }))
```

Reference: https://github.com/Cyfrin/aderyn/actions/runs/8772911977/job/24072381859?pr=333#step:6:792

**Comment by TilakMaddy** [comment]
Once we have tests working, we should try to replace all `load_contract` with `load_contract_directly` (i think)

**Comment by TilakMaddy** [comment]
Okay so now CI tests are all passing however I'm not sure if you re-run the jobs the `cargo test` will still pass. That's the only once causing issues which is why I have introduced a `take_load_contract_lock()` ...still not confident :P

**Comment by TilakMaddy** [comment]
@alexroan  Maybe from now on, we should make PRs to this branch, instead of directly pushing. This way we can see if the tests are passing before we merge into this.

**Comment by TilakMaddy** [comment]
https://github.com/foundry-rs/compilers/issues/113 Issue opened for help

**Comment by TilakMaddy** [comment]
@alexroan I have removed the test `test_siblings_by_loading_contract_directly` in order for CI to pass

**Comment by alexroan** [comment]
> @alexroan Maybe from now on, we should make PRs to this branch, instead of directly pushing. This way we can see if the tests are passing before we merge into this.

@TilakMaddy 100% that is the intention for this branch. I was merging directly from dev into this, but I'll open PRs for that too.

**Comment by TilakMaddy** [comment]
All detectors' tests now load contracts directly!

**Comment by TilakMaddy** [comment]
> @alexroan Maybe from now on, we should make PRs to this branch, instead of directly pushing. This way we can see if the tests are passing before we merge into this.

said @TilakMaddy and 2 hours later he proceeded to do just the opposite . . .

**Comment by TilakMaddy** [comment]
Plan

To explore next: **PoC for handling a simple remapping sample for imports in a contract within the function `load_solidity_source_unit_directly()`.** 

(I haven't started this exploration yet)

Then following
* Make that option available in 3 ways
 1. load_solidity_source_unit_directly
 2. load_solidity_source_unit_directly_with_dependencies
 3. load_solidity_source_unit_directly_with_dependencies_and_scope_exlcuded

(probably create an enum to make functions look clean)

* After all that's done and tested, integrate it with aderyn driver to make remapping available to the whole report generation process

**Comment by TilakMaddy** [comment]
The tests more or less seem to be stable now. We can say they aren't causing a problem except that we are forced to run them serially. Later we can hopefully find time to work on a testing framework built on foundry_compilers which can help with tests in 2 phases ->
 1. Generate ASTs (happens serially)
 2. detector.detect() (happens in parallel)

This can save a lot od time especially when we get to 400+ detectors

**Comment by alexroan** [comment]
Before you push anymore commits to this branch @TilakMaddy , let's catchup first. I want to understand what's already in here, and in future, all code changes here should be through PRs.

Breaking up the PRs means I can review them more easily, one-by-one. Reviewing the whole thing at the end with little context is going to introduce many bugs.

**Comment by TilakMaddy** [comment]
Yes @alexroan let's catch up first. The only reason why I pushed directly here after the message was because the tests were failing when they weren't supposed to according to the PR's `push` and `pull_request` status (It was passing). 

But you can see most of them were PRs #359 to #364  : P 100% we can breakdown PR by PR and review it together : )

I would say it's not too many broad changes actually. It was just me trying different things on the same lines (which is to make tests stable). Hence, so many commits

**Comment by alexroan** [comment]
[diff.txt](https://github.com/Cyfrin/aderyn/files/15147567/diff.txt)

This branch at commit 596d4f5 seems to exclude certain solidity files that weren't previously excluded. The diff attached is report.md diff from running `aderyn` v0.0.24, and this branch of aderyn on https://github.com/Cyfrin/aderyn-contracts-playground.

A couple of points:
* lib contracts are included (we already acknowledged this and it can be solved using `--exclude`)
* **A few contracts are not included**

Regarding the second point, the following contracts are missing:
* src/CrazyPragma.sol
* src/InconsistentUints.sol
* src/cloc/AnotherHeavilyCommentedContract.sol
* src/cloc/EmptyContractFile.sol

cc: @TilakMaddy

**Comment by alexroan** [comment]
Running `./cli/reportgen.sh` I get this error:

```
stderr = Error: Source "lib/openzeppelin-contracts/contracts/token/ERC721/ERC721.sol" not found: File not found. Searched the following locations: "".
 --> src/BasicNft.sol:5:1:
  |
5 | import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error: Source "lib/openzeppelin-contracts/contracts/token/ERC721/ERC721.sol" not found: File not found. Searched the following locations: "".
 --> src/Initializer.sol:5:1:
  |
5 | import "@oz/contracts/token/ERC721/ERC721.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


cwd = ./tests/foundry-nft-f23
[aderyn_driver/src/process_auto.rs:88:17] compilation_groups = {
    "0.8.25": [
        "/Users/alexroan/Documents/Cyfrin/aderyn/tests/foundry-nft-f23/src/BasicNft.sol",
        "/Users/alexroan/Documents/Cyfrin/aderyn/tests/foundry-nft-f23/src/Initializer.sol",
        "/Users/alexroan/Documents/Cyfrin/aderyn/tests/foundry-nft-f23/src/inner-core-modules/ICM.sol",
    ],
}
thread 'main' panicked at aderyn_driver/src/process_auto.rs:89:17:
Error running solc command
stack backtrace:
   0: rust_begin_unwind
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:647:5
   1: core::panicking::panic_fmt
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/panicking.rs:72:14
   2: aderyn_driver::process_auto::with_project_root_at
             at ./aderyn_driver/src/process_auto.rs:89:17
   3: aderyn_driver::driver::make_context
             at ./aderyn_driver/src/driver.rs:117:24
   4: aderyn_driver::driver::drive
             at ./aderyn_driver/src/driver.rs:25:22
   5: aderyn::main
             at ./aderyn/src/main.rs:214:9
   6: core::ops::function::FnOnce::call_once
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by alexroan** [comment]
https://github.com/Cyfrin/foundry-compilers

Official fork of foundry compilers @TilakMaddy :)

**Comment by TilakMaddy** [comment]
> EDITED:
> 
> > [diff.txt](https://github.com/Cyfrin/aderyn/files/15147567/diff.txt)
> > This branch at commit [596d4f5](https://github.com/Cyfrin/aderyn/commit/596d4f5e78256f71338b5607c8c2c19fa23130fe) seems to exclude certain solidity files that weren't previously excluded. The diff attached is report.md diff from running `aderyn` v0.0.24, and this branch of aderyn on https://github.com/Cyfrin/aderyn-contracts-playground.
> > A couple of points:
> > 
> > * lib contracts are included (we already acknowledged this and it can be solved using `--exclude`)
> > * **A few contracts are not included**
> > 
> > Regarding the second point, the following contracts are missing:
> > 
> > * src/CrazyPragma.sol
> > * src/InconsistentUints.sol
> > * src/cloc/AnotherHeavilyCommentedContract.sol
> > * src/cloc/EmptyContractFile.sol
> 
> cc: @TilakMaddy
> 
> UPDATE: Running this with `--exclude /lib/` works perfectly, except for the `EmptyContractFile.sol` not showing.
> 
> PS: I'm going to resolve conflicts from dev here ASAP.
> 
> PPS: Please pull the latest @TilakMaddy 👍

@alexroan  we don't see `EmptyContractFile.sol` but that is by design. It is because it doesn't produce an AST (just like any other non solidity file present in the directory).

**Comment by TilakMaddy** [comment]
Alex let me know what is in your mind for us to achieve feature parity with `dev` so we can continue work on this!

**Comment by TilakMaddy** [comment]
There are 2 breaking changes as highlighted in the PR description. How might we tackle them? Maybe add a `--with-sensible-defaults` option? and a warning message if `requirements` is absent ??

**Comment by alexroan** [comment]
I'm going to battle-test this over the next few hours and throw error reports in here.

In the meantime, @TilakMaddy , can you resolve these conflicts, please?

**Comment by alexroan** [comment]
# Battle Test

target repo: https://github.com/Cyfrin/aderyn-contracts-playground

## Observations

### 1. Failing on `solc` fail

Current behavior: Large scary warnings and error traces from solc, then fail.

Desired behaviour: bail out with a reasonable error. For example, if it can't find certain `lib` dependencies, which will be common unless `--exclude /lib/` is passed, then there should be a reasonable error explaining so instead of a wall of text.

### 2. Foundry project issues

Running in the root of a foundry project often results in the error mentioned in observation 1, unless `--exclude /lib/` is passed. A reasonable next step as an engineer is to think "I'll just run aderyn inside `src/`". This produces the following stack trace:

```
Resolving sources versions by graph ...
Compiling 15 files with Solc 0.8.19+commit.7dd6d404.Darwin.appleclang
Compiling 43 files with Solc 0.8.20+commit.a1b79de6.Darwin.appleclang
Compiling 5 files with Solc 0.8.25+commit.b61c2a91.Darwin.appleclang
thread 'thread '<unnamed><unnamed>' panicked at ' panicked at aderyn_driver/src/lib.rsaderyn_driver/src/lib.rs::29:60:
called `Option::unwrap()` on a `None` value
29:60:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: core::option::unwrap_failed
   4: aderyn_driver::passes_scope
   5: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
   6: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
   7: aderyn_driver::process_auto::with_project_root_at::{{closure}}
   8: rayon::iter::plumbing::Folder::consume_iter
   9: rayon::iter::plumbing::bridge_producer_consumer::helper
  10: rayon_core::join::join_context::{{closure}}
  11: <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute
  12: rayon_core::registry::WorkerThread::wait_until_cold
  13: rayon_core::registry::ThreadBuilder::run
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: core::option::unwrap_failed
   4: aderyn_driver::passes_scope
   5: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
   6: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
   7: aderyn_driver::process_auto::with_project_root_at::{{closure}}
   8: rayon::iter::plumbing::Folder::consume_iter
   9: rayon::iter::plumbing::bridge_producer_consumer::helper
  10: rayon_core::join::join_context::{{closure}}
  11: rayon_core::registry::in_worker
  12: rayon::iter::plumbing::bridge_producer_consumer::helper
  13: <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute
  14: rayon_core::registry::WorkerThread::wait_until_cold
  15: rayon_core::registry::ThreadBuilder::run
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

We can probably handle `passes_scope` a bit better to help this, as that's where it's failing.
We should try to maintain the ability to just run `aderyn` in a foundry root. This may mean that we need to:
* alter --scope to account for relative paths, or have some other parameter for this
* "cheat" by peeping into the foundry.toml to get scope/exclude lists

### 3. Running on a single file

aderyn should support running on a single file, for example:

```
aderyn ./src/AdminContract.sol
``

Currently, this returns an error.

## Conclusion

This has come really far @TilakMaddy , excellent stuff. The challenges we're facing now are DevRel challenges. Most of the backend work is done, so lets try to put ourselves in the shoes of developers and auditors, and understand how they would intuitively _try_ to use the tool and cater to that.

**Comment by alexroan** [comment]
OK, Next Steps:
- `--icf` flag to run this framework
  - maintain user experience currently on the `dev` branch when this flag is not supplied
- `--scope` meaning change (`--scope ./src` for example)
  - Add another flag called `filter` to take existing `scope` functionality

Ping me when these are ready and we'll talk about merging!

**Comment by TilakMaddy** [comment]
@alexroan There is an error in the way I am reading remappings (Right now, it requires that you're remappings are placed in a certain order for this thing to work)

For example, consider this

```
@a/=lib/c
@a/b=lib/a/b
```

Now in my solidity files when I say  `import @a/b/f` 

We expect `lib/a/b/f` but it resolves to `lib/c/b/f`

The current fix is to flip the order in the remappings

```
@a/b=lib/a/b
@a/=lib/c
```

Now the first pattern will be matched first and it works !!

IMO this doesn't affect when remappings are generated and given to us like `forge remappings > remappings` but it can cause problems when people manually write their remappings.

So I don't know how big a deal this ... just letting you know (we can work on it)

**Comment by TilakMaddy** [comment]
2 more things that we're going to do before moving to dev with `--icf` flag:

* Single file support
* Auditor mode's found instances

**Comment by TilakMaddy** [comment]
Blocker #413, #414

**Comment by TilakMaddy** [comment]
I think we do have single-file support in a way.
`cargo run ./tests/contract-playground --scope SingleFile.sol`

**Comment by TilakMaddy** [comment]
@alexroan Can you work on approving PRs in the range #413 to #425 (in reverse order) ... So that when it's done we can have a clean slate with full feature parity that can help us plan our next steps  (which is introducing `--icf` flag and merge with dev)

**Comment by TilakMaddy** [comment]
It looks like `pull_request` CI errors are because we haven't merged the latest `dev`

**Comment by TilakMaddy** [comment]
@alexroan  Please run `git reset --hard HEAD~1` and then `git push --force` .

**Comment by TilakMaddy** [comment]
@alexroan FYI There is a separate PR #433 being worked on where we bring CI tests of the existing dev branch into ICF here!!

So please let us not alter the workflow file here (until we complete that)

**Comment by TilakMaddy** [comment]
Now, that we have sarif and auditor mode working here,  let me work on the PR which gives us the ICF flag 

After that we can #GoToDEV !! 🙌

**Comment by TilakMaddy** [comment]
UPDATE: Now we have `--icf` flag without that it will switch to the current dev behavior. (so its fully backward compatible)

However, I need to work on CI tests (to confirm the above)  

(That's a separate PR - stay tuned 😎)

So far checkpoint is that we can confirm that all commands with `--icf` work as expected : )

**Comment by TilakMaddy** [comment]
Now, this looks like it can make it to dev  🚀   🤞

**Comment by TilakMaddy** [comment]
There are a few things on my mind I still want to do before we do _The Merge_  : 

1) Use config crate to resolve scope and exclude for legacy framework too (not just ICF which is what we have) !

**Comment by TilakMaddy** [comment]
Run `make setup` after you pull this branch !

**Comment by TilakMaddy** [comment]
Oops, I might have made a mistake on sarif report generation. Let me crosscheck it. It may require more work that I thought.

On hold until then

**Comment by TilakMaddy** [comment]
Yooo I FOUND OUT exactly 🤯 why sarif reports look different in dev than they look here !
AND it had noting to do with ICF itself

**Comment by TilakMaddy** [comment]
TODO: Single file support

File: aderyn_core/src/detect/high/block_timestamp_deadline.rs [hunk 1]
```diff
@@ -116,55 +116,23 @@ impl IssueDetector for BlockTimestampDeadlineDetector {
...
 #[cfg(test)]
 mod block_timestamp_deadline_detector_tests {
-    use crate::detect::{
-        detector::{detector_test_helpers::load_contract, IssueDetector},
-        high::BlockTimestampDeadlineDetector,
-    };
+    use serial_test::serial;
 
-    #[test]
-    fn test_block_timestamp_deadline_uniswap_v2_detector() {
```
Comment (alexroan) [inline aderyn_core/src/detect/high/block_timestamp_deadline.rs:125]: Can you readd this test please?

We want to test both the UniswapV2 and UniswapV3 contracts

File: aderyn_core/src/detect/test_utils/load_contract.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+use crate::context::workspace_context::WorkspaceContext;
+use crate::framework::foundry::read_foundry_output_file;
+use crate::read_file_to_string;
+use crate::visitor::ast_visitor::Node;
+use std::path::PathBuf;
+
+pub fn load_contract(filepath: &str) -> WorkspaceContext {
```
Comment (alexroan) [inline aderyn_core/src/detect/test_utils/load_contract.rs:7]: I think we can remove this file altogether. The only files using this function are _template.rs!

File: aderyn_core/src/detect/test_utils/load_source_unit.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        eprintln!("Error running solc command");
+        std::process::exit(1);
+    }
+}
+
+#[cfg(test)]
+/// This function is dangerous to use because we force all the sol files into 1 Workspace Context.
+/// As a result, we may override Node IDs. Therefore, this function is only available in cfg(test)
+#[allow(dead_code)]
+pub fn load_multiple_solidity_source_units_into_single_context(
```
Comment (alexroan) [inline aderyn_core/src/detect/test_utils/load_source_unit.rs:83]: Can you write a couple of tests for this?

The case I want to cover is different solc versions to be loaded into the same context. We should not allow that if it is currently possible.

File: aderyn_core/src/framework/foundry.rs [hunk 4]
```diff
@@ -209,29 +210,3 @@ fn get_matching_output_files(
         .cloned()
         .collect()
 }
-
```
Comment (alexroan) [inline aderyn_core/src/framework/foundry.rs:212]: Reason for removing this one?

File: aderyn_driver/src/process_foundry.rs [hunk 5]
```diff
@@ -10,12 +10,13 @@ use std::{
     path::{Path, PathBuf},
 };
 
+#[allow(dead_code)]
```
Comment (alexroan) [inline aderyn_driver/src/process_foundry.rs:13]: Do we need this?

**Review by alexroan — COMMENTED** [review]
A few comments to resolve @TilakMaddy .

We are _very_ close to merging this.

File: aderyn_core/src/framework/foundry.rs (hunk 4 above)
Comment (alexroan) [reply aderyn_core/src/framework/foundry.rs:212]: I think this is fine, it seems to test foundry more than aderyn's interpretation of foundry output files

File: .github/workflows/cargo.yml [hunk 6]
```diff
@@ -63,41 +93,100 @@ jobs:
...
+          cd ../../../
+
       # Verify report.md
 
       - name: Generate report-workflow.md
         uses: actions-rs/cargo@v1
         with:
           command: run
-          args: -- -o ./reports/report-workflow.md ./tests/contract-playground/ --skip-update-check
+          args: -- -o ./reports/report-workflow.md  --src src/ ./tests/contract-playground/ --skip-update-check --icf
```
[the remark this answers is not in the document]
Comment (alexroan) [reply .github/workflows/cargo.yml:120]: Let's leave this for now, but worth putting a pin in for future work on `dev`

File: aderyn_core/src/detect/test_utils/load_source_unit.rs (hunk 3 above)
Comment (alexroan) [reply aderyn_core/src/detect/test_utils/load_source_unit.rs:82]: On second thoughts, I'm keen to merge this to dev, so this is another one I think we should follow up on later @TilakMaddy .

**Pull request merged**